cranelift-codegen/src/ir/constant.rs

```diff
@@ -52,6 +57,72 @@ impl From<V128Imm> for ConstantData {
     }
 }
 
+macro_rules! try_into_int {
+    ($type: ty) => {
+        impl TryInto<$type> for ConstantData {
+            type Error = String;
+
+            fn try_into(self) -> Result<$type, Self::Error> {
+                if self.len() == (mem::size_of::<$type>()) {
+                    let v = self.into_vec().into_iter().rev();
+                    let mut r: $type = 0;
+                    for b in v {
+                        let (mut shifted, _) = r.overflowing_shl(8);
+                        shifted |= b as $type;
+                        r = shifted;
+                    }
+                    Ok(r)
+                } else {
+                    Err(format!(
+                        "Incorrect vector size: {}, expected {}",
+                        self.len(),
+                        mem::size_of::<$type>()
+                    ))
+                }
+            }
+        }
+    };
+}
+
+try_into_int!(u8);
+try_into_int!(u16);
+try_into_int!(u32);
+try_into_int!(u64);
+try_into_int!(u128);
+
+impl TryInto<f32> for ConstantData {
+    type Error = String;
+
+    fn try_into(self) -> Result<f32, Self::Error> {
+        let bits = self.try_into();
+        match bits {
+            Ok(v) => Ok(f32::from_bits(v)),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl TryInto<f64> for ConstantData {
+    type Error = String;
+
+    fn try_into(self) -> Result<f64, Self::Error> {
+        let bits = self.try_into();
+        match bits {
+            Ok(v) => Ok(f64::from_bits(v)),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl TryInto<bool> for ConstantData {
+    type Error = String;
+
+    fn try_into(self) -> Result<bool, Self::Error> {
+        // Only check the least significant byte regardless of size.
+        Ok(self.into_vec()[0] != 0)
+    }
+}
+
```

Comment:
I don't think we should modify `ConstantData` for what we are doing; it is designed for constant pool stuff and adding these conversions might make it harder to decouple later from test functionality. Why not build our own `Value` enum:
```
enum Value {
  Bool(bool),
  F32(f32),
  F64(f64),
  ... 
}
```
I wouldn't worry too much about 128-bit types just yet until we get the basics in place. Then I would add a way to parse into this (using Rust's `FromStr/parse`) from text and a type:
```
impl Value {
  fn parse_as(text: &str, type: Type) -> Result<Self, ...> {
    match type {
      B8 | B16 | ... => Ok(Value::Bool(text.parse()?)),
      F32 => Ok(Value::F32(text.parse()?)),
      ...
}
```
Then I would `#[derive(Eq, PartialEq)]` on the enum so we can compare them to each other and add some `From` implementations so that `invoke` can get the raw value (e.g. f32), wrap it in a `Value` with `Value::from` (or `into`), and then compare it against the parsed `Value`s that we get from the comments. (Hopefully that explains the other comments I've made; this isn't the only way to do things but it doesn't involve touching other components like `ConstantData` and it seems natural to Rust--from the little I know about Rust).

Reply:
Let me change my tune a bit: I now see down below that you implemented some parsing logic in `parser.rs`; if you want to use that for the `FromStr` on `Value` instead of the built-in Rust parsing that seems OK. We could use methods like `Parser::new(text).match_bool(...)` instead and `Value` may have to use different inner types (e.g. `F32(Ieee32)` instead of `F32(f32)`)--but that seems fine. The underlying thing that I'm getting at is that we likely don't want to touch `ConstantData` in this case.